### PR TITLE
Add support for deserializing unordered arrays

### PR DESCRIPTION
**Pull request by reyk**
This PR does two things:

1. It fixes deserializing maps with integers as keys
2. It adds a helper `deserialize_unordered_array` to deserialize PHP arrays where the indexes are out of order.

To small changes are included to make clippy happy.

### CODE REVIEW

File: src/de.rs [hunk 1]
```diff
@@ -523,9 +533,22 @@ where
...
+pub fn deserialize_unordered_array<'de, T, D>(
+    deserializer: D,
+) -> std::result::Result<Vec<T>, D::Error>
+where
+    D: Deserializer<'de>,
+    T: Deserialize<'de>,
+{
+    // Serialize into a map and return a Vec ordered by the keys.
+    let v = BTreeMap::<usize, T>::deserialize(deserializer)?;
+    Ok(v.into_iter().map(|(_, v)| v).collect())
```
Comment (mbr) [inline src/de.rs:561]: Am I reading this incorrectly, or does this not discard "holes"?

I.e. what happens (or should happen) to the following PHP array:

```php
$arr = array();
$arr[0] = "zero";
$arr[2] = "two";
$arr[1] = "one";
$arr[6] = "six";
```

Would this not come out as a length four array "one", "two", "three", "six"? Do we really want that?

If we do, this should at least be reflected in the function name and docs, and have a test case for this edge case.

**Review by mbr — COMMENTED** [review]
Question about the helper function.

File: src/de.rs (hunk 1 above)
Comment (reyk) [reply src/de.rs:561]: It would come out as a length 4 array indeed, dropping the holes. It should be documented  and I have a test case. But what should we do otherwise, I don‘t think that we can fill in the empty fields (with what?).

Comment (reyk) [reply src/de.rs:561]: I added an extra commit to document it. Let me know if you have a better idea or wording.
